storage/engine/mvcc.go

```diff
@@ -985,70 +1025,57 @@ func MVCCDeleteRange(engine Engine, ms *MVCCStats, key, endKey roachpb.Key, max
 	return num, nil
 }
 
-func getScanMeta(iter Iterator, encEndKey MVCCKey, meta *MVCCMetadata) (roachpb.Key, MVCCKey, error) {
+func getScanMeta(iter Iterator, encEndKey MVCCKey, meta *MVCCMetadata) (MVCCKey, error) {
 	metaKey := iter.Key()
-	if bytes.Compare(metaKey, encEndKey) >= 0 {
-		return nil, nil, iter.Error()
-	}
-	key, ts, isValue, err := MVCCDecodeKey(metaKey)
-	if err != nil {
-		return nil, nil, err
+	if !metaKey.Less(encEndKey) {
+		return NilKey, iter.Error()
 	}
-	if isValue {
+	if metaKey.IsValue() {
 		meta.Reset()
-		meta.Timestamp = ts
-		meta.KeyBytes = int64(len(iter.unsafeKey()))
+		meta.Timestamp = metaKey.Timestamp
+		meta.KeyBytes = mvccVersionTimestampSize
 		meta.ValBytes = int64(len(iter.unsafeValue()))
 		meta.Deleted = len(iter.unsafeValue()) == 0
-		metaKey = metaKey[:len(metaKey)-int(mvccVersionTimestampSize)]
-		return key, metaKey, nil
+		return metaKey, nil
 	}
 	if err := iter.ValueProto(meta); err != nil {
-		return nil, nil, err
+		return NilKey, err
 	}
-	return key, metaKey, nil
+	return metaKey, nil
 }
 
-func getReverseScanMeta(iter Iterator, encEndKey MVCCKey, meta *MVCCMetadata) (roachpb.Key, MVCCKey, error) {
+func getReverseScanMeta(iter Iterator, encEndKey MVCCKey, meta *MVCCMetadata) (MVCCKey, error) {
 	metaKey := iter.Key()
 	// The metaKey < encEndKey is exceeding the boundary.
-	if bytes.Compare(metaKey, encEndKey) < 0 {
-		return nil, nil, iter.Error()
+	if metaKey.Less(encEndKey) {
+		return NilKey, iter.Error()
 	}
 
-	// The row with oldest version will be got by seeking reversely. We use the
-	// key of this row to get the MVCC metadata key.
-	key, _, isValue, err := MVCCDecodeKey(metaKey)
-	if err != nil {
-		return nil, nil, err
-	}
 	// If this isn't the meta key yet, scan again to get the meta key.
 	// TODO(tschottdorf): can we save any work here or leverage
 	// getScanMetaKey() above after doing the Seek() below?
-	if isValue {
-		iter.Seek(MVCCEncodeKey(key))
+	if metaKey.IsValue() {
+		// The row with oldest version will be got by seeking reversely. We use the
+		// key of this row to get the MVCC metadata key.
+		iter.Seek(MakeMVCCKey(metaKey.Key))
```

Comment:
This line shows that `MakeMVCCKey` should probably be called `MakeMVCCMetadataKey`.